### PR TITLE
More gates supported natively

### PR DESCRIPTION
**Pull request by lducas**
- added support for Homomorphic NOR, OR, AND and NOT
- updated test program to include all gates
- added checks of independence for input ciphertexts to avoid buggy use:
https://github.com/lducas/FHEW/issues/13

### CODE REVIEW

**Comment by lducas** [comment]
Apologies for the ``white-space'' noise. My editor is not so clever and has trouble with for-loops without brackets.

**Comment by micciancio** [comment]
Gates look fine, can be merged as it is. Comments below are mostly for readability, and can be addressed later.

HomGate: condition on dependent ciphertexts includes also ciphertexts that are neither equal not one the negation of the other. (Each a[i] can be negated independently.) No need to change, as condition is a conservative approximation, and false positives have exponentially small probability.
Also, "+q/4" term in call to InitializeACC could be moved to GateConst table. 

fhewTest: very hard to follow, both code and output. Fine for testing, but not a good example on how to use library.
